src/main/java/uk/nhs/tis/trainee/actions/service/ActionService.java

```diff
@@ -58,12 +64,41 @@ public class ActionService {
    * The constructor of action service.
    */
   public ActionService(ActionRepository repository, ActionMapper mapper,
-                       EventPublishingService eventPublishingService) {
+      EventPublishingService eventPublishingService) {
     this.repository = repository;
     this.mapper = mapper;
     this.eventPublishingService = eventPublishingService;
   }
 
+  /**
+   * Add or update actions for a given Placement DTO.
+   *
+   * @param dto     The placement DTO.
+   * @param actions The list of actions to supplement with new or updated actions.
+   */
+  private void addOrUpdatePlacementAction(PlacementDto dto, List<Action> actions) {
```

Comment:
This was moved out of the `updateActions` for placements to reduce cognitive load. I had thought to see if it was possible to refactor out the common logic between programme, placement and person actions, but it was looking like being even more complicated so I didn't progress that thought.